### ISSUE TITLE
Are you open to some build and project layout refactoring?

### ISSUE THREAD

**Issue by trevjonez** [sonatype-nexus-community/scan-gradle-plugin#32]
In order to support some changes that I would like to make to resolve in #31 (where I would add explicit support for android projects,) I would like to make some adjustments to the project build and layout.

Is this project and maintainers open to such a contribution?

Specifically I would like to restructure the top level to be a wrapper on other projects, hopefully simplifying the way you are having to consume the nexus-platform-api. most likely by deferring the shadowing until deployment of a specific configuration of the gradle plugin module and enabling it to consume it using standard gradle conventions.( ie not via direct reference to the output jar path since it ends up depending on implementation details of that tool chain) 

This would hopefully lower ongoing maintenance effort in the long run by letting gradle handle the harder work. and open the doors to a more modular project layout without duplicating configuration in the process

Then to actually resolve #31, I would be able to make a gradle module specific to android so that builds using this plugin that are not android are not forced to pull in those dependencies. At that point this project would produce two distinct entries to the plugin portal and the android project could delegate to the existing project when appropriate.

Additionally this could enable publishing of two variants of the main plugin. one with shadowing and one without. again the goal being to minimize transitive dependency downloads where the standard gradle resolution would actually be sufficient. (though perhaps that is not possible and I am lacking sufficient background on an issue forcing the project to do it the way it already is?)

TLDR: if I do the work to change the project layout/build to clear the path for future enhancements, is the project receptive to this type of change?

**Comment by guillermo-varela** [comment]
Hi @trevjonez,

The project is totally open to take contributions from community. Feel free to fork, develop and submit your PRs.

Please include unit and integration tests so we can validate the changes, specially if you're about to send a big one.

Thanks.

**Comment by guillermo-varela** [comment]
If you decide to work on #31, please leave a comment there so we can now the issue is being addressed.

**Comment by guillermo-varela** [comment]
> Additionally this could enable publishing of two variants of the main plugin. one with shadowing and one without. again the goal being to minimize transitive dependency downloads where the standard gradle resolution would actually be sufficient. (though perhaps that is not possible and I am lacking sufficient background on an issue forcing the project to do it the way it already is?)

Shadowing (package relocation) is being done because `nexus-platform-api` is published as a fat/uber jar with all its dependencies inside, but without package relocation, and some users have reported classpath conflicts with those inner libraries. We can't modify how `nexus-platform-api` is published so the workaround had to be done directly in this plugin.

**Comment by guillermo-varela** [comment]
> Additionally this could enable publishing of two variants of the main plugin.

@trevjonez one comment about that: `nexus-platform-api` is only required for the Nexus IQ Server integration, not for OSS Index. But a decision was made to make Sonatype plugins supporting both platforms, so dividing this into two different plugins is not something we're huge fans of, unless there is a big enough reason to do so. That's why we have a single plugin with separated tasks per platform.

And final tip I can give you is to take a look at: https://github.com/sonatype-nexus-community/scan-gradle-plugin/issues/29

Apparently `configuration.getResolvedArtifacts()` has [a reported issue at the Gradle repo](https://github.com/gradle/gradle/issues/5426) regarding Android (I'm not sure if that was one of things you were planning to change). The reason we use that method is because Nexus IQ Server needs the path of the artifacts to scan the dependencies binaries and perform a more accurate analysis than using just each dependency coordinates (group, name and version). That means access to the resolved artifact files is required for both "regular" Gradle projects and Android projects as well.

**Comment by trevjonez** [comment]
Thanks for the insight on the history of things. I have not had a chance to dig in any more than the original issue description so plans are still vague at best. In general I like to avoid shadowing whenever possible but it seems like you have a reasonable case for it here.

as for configuration artifact resolution, that is one I have not been up against before so I'll be sure to look a bit closer at that. I imagine there is something doable to disambiguate between the artifacts produced from an android build. Best guess in my mind right now is some sort of filter operation on the live collections types that gradle uses throughout the public API's.

I am hoping to start getting into this in the next week or so.

**Comment by trevjonez** [comment]
looks like some good progress has been made toward furthering android support which was my main motivator for this ticket. Perhaps i'll revisit in the future but no point in keeping this stale ticket up for now.

Thanks all for the efforts here!
